### PR TITLE
`plonk::Error::TableError`: Better lookup errors

### PR DESCRIPTION
**Pull request by therealyingtong**
"All good lookups are alike, but every bad lookup is bad in its own way" - @nalinbhardwaj

Supersedes #750, inspired by https://github.com/zcash/halo2/pull/750#discussion_r1141000039.

### CODE REVIEW

File: halo2_proofs/src/circuit/layouter.rs [hunk 1]
```diff
@@ -109,24 +109,6 @@ pub trait RegionLayouter<F: Field>: fmt::Debug {
     fn constrain_equal(&mut self, left: Cell, right: Cell) -> Result<(), Error>;
 }
 
-/// Helper trait for implementing a custom [`Layouter`].
-///
-/// This trait is used for implementing table assignments.
-///
-/// [`Layouter`]: super::Layouter
-pub trait TableLayouter<F: Field>: fmt::Debug {
```
Comment (daira) [inline halo2_proofs/src/circuit/layouter.rs:117]: Let's re-export this here to reduce API breakage.

File: halo2_proofs/src/circuit.rs [hunk 2]
```diff
@@ -13,6 +13,9 @@ pub mod floor_planner;
 pub use floor_planner::single_pass::SimpleFloorPlanner;
 
 pub mod layouter;
+pub mod table_layouter;
```
Comment (ebfull) [inline halo2_proofs/src/circuit.rs:16]: Let's keep this module private, so just `mod` I think.

**Review by ebfull — CHANGES_REQUESTED** [review]

**Review by ebfull — APPROVED** [review]

File: halo2_proofs/src/circuit/table_layouter.rs [hunk 3]
```diff
@@ -106,23 +115,293 @@ impl<'r, 'a, F: Field, CS: Assignment<F> + 'a> TableLayouter<F>
...
             if assigned.iter().all(|b| *b) {
-                Some(assigned.len())
+                // All values in the column have been assigned
+                Ok((col, assigned.len()))
+            } else {
+                Err(Error::TableError(TableError::ColumnNotAssigned(*col)))
+            }
+        })
+        .collect();
+    let column_lengths = column_lengths?;
```
Comment (daira) [inline halo2_proofs/src/circuit/table_layouter.rs:135]:
```suggestion
        .collect()?;
```

File: halo2_proofs/src/circuit/table_layouter.rs [hunk 4]
```diff
@@ -106,23 +115,293 @@ impl<'r, 'a, F: Field, CS: Assignment<F> + 'a> TableLayouter<F>
...
 }
 
-pub(crate) fn compute_table_lengths<F>(
+pub(crate) fn compute_table_lengths<F: Debug>(
     default_and_assigned: &HashMap<TableColumn, (DefaultTableValue<F>, Vec<bool>)>,
 ) -> Result<usize, Error> {
-    match default_and_assigned
-        .values()
-        .map(|(_, assigned)| {
+    let column_lengths: Result<Vec<_>, Error> = default_and_assigned
```
Comment (daira) [inline halo2_proofs/src/circuit/table_layouter.rs:121]:
```suggestion
    let column_lengths = default_and_assigned
```

**Review by daira — APPROVED** [review]
utACK

**Pull request merged**